Derivs/src/derivs_spacetimex.hxx

```diff
@@ -0,0 +1,512 @@
+    //          dx;
+    // else
+    //   return (-1 / T(4) * var[-di]      //
+    //           - 5 / T(6) * var[0]       //
+    //           + 3 / T(2) * var[+di]     //
+    //           - 1 / T(2) * var[+2 * di] //
+    //           + 1 / T(12) * var[+3 * di]) /
+    //          dx;
+
+    const simd<T> symm =
+        -1 / T(24) *
+            (maskz_loadu(mask, &var[-3 * di]) -
+             maskz_loadu(mask, &var[3 * di])) //
+        + 1 / T(4) *
+              (maskz_loadu(mask, &var[-2 * di]) -
+               maskz_loadu(mask, &var[2 * di])) //
+        -
+        7 / T(8) * (maskz_loadu(mask, &var[-di]) - maskz_loadu(mask, &var[di]));
+    const simd<T> anti =
+        -1 / T(24) *
+            (maskz_loadu(mask, &var[-3 * di]) +
+             maskz_loadu(mask, &var[3 * di])) //
+        + 1 / T(4) *
+              (maskz_loadu(mask, &var[-2 * di]) +
+               maskz_loadu(mask, &var[2 * di])) //
+        - 5 / T(8) *
+              (maskz_loadu(mask, &var[-di]) + maskz_loadu(mask, &var[di])) //
+        + 5 / T(6) * maskz_loadu(mask, &var[0]);
+    return (vel * symm - fabs(vel) * anti) / dx;
+  }
+}
+
+template <int dir, typename T, int D>
+inline ARITH_INLINE ARITH_DEVICE ARITH_HOST simd<T>
+deriv_upwind(const simdl<T> &mask, const GF3D2<const T> &gf_,
+             const vect<int, dim> &I, const vec<simd<T>, D> &vel,
+             const vec<T, D> &dx) {
+  static_assert(dir >= 0 && dir < D, "");
+  const auto &DI = vect<int, dim>::unit;
+  const ptrdiff_t di = gf_.delta(DI(dir));
+  return deriv1d_upwind(mask, &gf_(I), di, vel(dir), dx(dir));
+}
+
+template <typename T>
+inline ARITH_INLINE ARITH_DEVICE ARITH_HOST simd<T>
+deriv_upwind(const simdl<T> &mask, const GF3D2<const T> &gf_,
+             const vect<int, dim> &I, const vec<simd<T>, dim> &vel,
+             const vec<T, dim> &dx) {
+  return deriv_upwind<0>(mask, gf_, I, vel, dx) +
+         deriv_upwind<1>(mask, gf_, I, vel, dx) +
+         deriv_upwind<2>(mask, gf_, I, vel, dx);
+}
+
+template <typename T>
+inline ARITH_INLINE ARITH_DEVICE ARITH_HOST simd<T>
+deriv1d_diss(const simdl<T> &mask, const T *restrict const var,
+             const ptrdiff_t di, const T dx) {
+  if constexpr (deriv_order == 2)
+    return ((maskz_loadu(mask, &var[-2 * di]) +
+             maskz_loadu(mask, &var[+2 * di])) //
+            -
+            4 * (maskz_loadu(mask, &var[-di]) + maskz_loadu(mask, &var[+di])) //
+            + 6 * maskz_loadu(mask, &var[0])) /
+           dx;
+  if constexpr (deriv_order == 4)
+    return ((maskz_loadu(mask, &var[-3 * di]) +
+             maskz_loadu(mask, &var[+3 * di])) //
+            - 6 * (maskz_loadu(mask, &var[-2 * di]) +
+                   maskz_loadu(mask, &var[+2 * di])) //
+            + 15 * (maskz_loadu(mask, &var[-di]) +
+                    maskz_loadu(mask, &var[+di])) //
+            - 20 * maskz_loadu(mask, &var[0])) /
+           dx;
+}
+
+using Arith::pown;
+
+template <int dir, typename T, int D>
```

Comment:
The functions above assume a fixed number of dimensions. This function is now generic (with a `D` template parameter). It also mixes `dim` and `D`. Is `D` necessary?